Sources/BraintreeCard/BTCardClient.m

```diff
@@ -133,34 +141,34 @@ - (void)tokenizeCard:(BTCard *)card completion:(void (^)(BTCardNonce *tokenizedC
                          }
                      }
                      
-                     [self sendAnalyticsEventWithSuccess:NO];
-
-                     completion(nil, callbackError);
+                     [self notifyFailure:callbackError completion:completion];
                      return;
                  }
-
+                
                  BTJSON *cardJSON = body[@"creditCards"][0];
+                 if (cardJSON.asError) {
+                    [self notifyFailure:cardJSON.asError completion:completion];
+                    return;
+                 }
                 
-                 [self sendAnalyticsEventWithSuccess:!cardJSON.isError];
-
-                 // cardNonceWithJSON returns nil when cardJSON is nil, cardJSON.asError is nil when cardJSON is non-nil
                  BTCardNonce *cardNonce = [BTCardNonce cardNonceWithJSON:cardJSON];
-                 completion(cardNonce, cardJSON.asError);
+                 [self notifySuccess:cardNonce completion:completion];
+                 return;
              }];
         }
     }];
 }
 
-#pragma mark - Analytics
+#pragma mark - Notify Result to Merchant
 
-- (void)sendAnalyticsEventWithSuccess:(BOOL)success {
-    NSString *event = [NSString stringWithFormat:@"ios.%@.card.%@", self.apiClient.metadata.integrationString, success ? @"succeeded" : @"failed"];
-    [self.apiClient sendAnalyticsEvent:event];
+- (void)notifyFailure:(NSError *)error completion:(void (^)(BTCardNonce *tokenizedCard, NSError *error))completion {
```

Comment:
Maybe `callBackFailure`? Duplicating the event string may even be an option here; or refactoring the internal logic of the method so that the completion handler is only invoked once near the end.

Reply:
Yeh - maybe the simplest option is just to do what we've always done and duplicate the event string everywhere we call the completion?

Another idea is to have a file similar to our `<Module>Error.swift` files (per module) that houses all analytics strings for a module. This way all our analytic strings are in one place, and we can have our 👀 more easily on all the naming conventions, as well as have a test file specifically for the event names.

Something like:

```
func paymentRequest() {
    apiClient.sendAnalyticsEvent(ApplePayAnalyticEvents.paymentRequestStarted)
}
....

func tokenize() {
     ...
    if error {
        apiClient.sendAnalyticsEvent(ApplePayAnalyticEvents.tokenizeFailed)
        completion(nil, error)
    }
}
```

Reply:
I think I like the second idea best. Then if there is change in logic and type of error added or changed, we can modify the constants in one place. I remember seeing this pattern elsewhere.

Reply:
I dig the latter idea, would also allow us to use the completion as is vs passing through to a new completion that handles the analytics. I think we should run with this as it’s the best of all of the worlds!

Reply:
We implemented this here: https://github.com/braintree/braintree_ios/pull/959/files. Once we are 👍 on the patter we can update the other 2 open PRs (card and amex) to follow the same.